tests/Main.hs

```diff
@@ -28,6 +31,26 @@ test6 _ _ = natVal (Proxy :: Proxy ((n^m)+(n*m)))
 test11 :: forall m . (KnownNat m) => Proxy m -> Integer
 test11 _ = natVal (Proxy @ (m*m))
 
+test12 :: forall m . (KnownNat (m+1)) => Proxy m -> Integer
+test12 = natVal
+
+test13 :: forall m . (KnownNat (m+3)) => Proxy m -> Integer
+test13 = natVal
+
+test14 :: forall m . (KnownNat (4+m)) => Proxy (7+m) -> Integer
+test14 = natVal
+
+type family Foo (m :: Nat) = (result :: Nat) | result -> m
```

Comment:
I have just tried this. Unfortunately these two testcases
```
    KnownNat (4 + Foo 1) => KnownNat (Foo 1 + 7); @ Foo 1 ~ 1
    KnownNat (4 + Foo 1 + Foo 1) => KnownNat (Foo 1 + 7 + Foo 1); @ Foo 1 ~ 1
```
are not triggered when I follow your suggestion. I am keeping things as written.